modules/pam_pwhistory/pwhistory_helper.c

```diff
@@ -0,0 +1,208 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote
+ *    products derived from this software without specific prior
+ *    written permission.
+ *
+ * ALTERNATIVELY, this product may be distributed under the terms of
+ * the GNU Public License, in which case the provisions of the GPL are
+ * required INSTEAD OF the above restrictions.  (This clause is
+ * necessary due to a potential bad interaction between the GPL and
+ * the restrictions contained in a BSD-style copyright.)
+ *
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <syslog.h>
+#include <errno.h>
+#include <unistd.h>
+#include <signal.h>
+#include <security/_pam_types.h>
+#include <security/_pam_macros.h>
+#include "opasswd.h"
+
+#define MAXPASS 200
+
+static void
+su_sighandler(int sig)
+{
+#ifndef SA_RESETHAND
+        /* emulate the behaviour of the SA_RESETHAND flag */
+        if ( sig == SIGILL || sig == SIGTRAP || sig == SIGBUS || sig = SIGSERV ) {
+		struct sigaction sa;
+		memset(&sa, '\0', sizeof(sa));
+		sa.sa_handler = SIG_DFL;
+                sigaction(sig, &sa, NULL);
+	}
+#endif
+        if (sig > 0) {
+                _exit(sig);
+        }
+}
+
+static void
+setup_signals(void)
+{
+  struct sigaction action;        /* posix signal structure */
+
+  /*
+   * Setup signal handlers
+   */
+  (void) memset((void *) &action, 0, sizeof(action));
+  action.sa_handler = su_sighandler;
+#ifdef SA_RESETHAND
+  action.sa_flags = SA_RESETHAND;
+#endif
+  (void) sigaction(SIGILL, &action, NULL);
+  (void) sigaction(SIGTRAP, &action, NULL);
+  (void) sigaction(SIGBUS, &action, NULL);
+  (void) sigaction(SIGSEGV, &action, NULL);
+  action.sa_handler = SIG_IGN;
+  action.sa_flags = 0;
+  (void) sigaction(SIGTERM, &action, NULL);
+  (void) sigaction(SIGHUP, &action, NULL);
+  (void) sigaction(SIGINT, &action, NULL);
+  (void) sigaction(SIGQUIT, &action, NULL);
+}
+
+static int
+read_passwords(int fd, int npass, char **passwords)
```

Comment:
Do you really need this function?  Isn't pam_modutil_read enough?

Reply:
I think it was wrong to replace read_passwords with pam_modutil_read. The return value is different. If that is fixed, it will probably somewhat work but the read_passwords handles reads from pipe better because it does not try to read until eof but just until it encounters the `\0` character. We have the same function in the pam_unix helpers.

Reply:
pam_modutil_read has a different semantics of the return value, and it doesn't null-terminate.
However, I still think read_passwords is overkill here: we don't need multiple passwords, and terminating the loop on reading '\0' is not something we could benefit from.

Reply:
Why we would not benefit?

Reply:
Do you want to impose a requirement that the data passed through this pipe must be null-terminated?

Reply:
The read_passwords handled that properly. And yes, that requirement was there and is there for the other helpers as well. Anything that is passed and is not null terminated is ignored.

Reply:
BTW, why this implementation of read_passwords differs from the one defined in modules/pam_unix/passverify.c?
For example,
```
-        /* The passwords array must contain npass preallocated
-         * buffers of length MAXPASS + 1
-         */
         int rbytes = 0;
         int offset = 0;
         int i = 0;
         char *pptr;
-        while (npass > 0) {
-                rbytes = read(fd, passwords[i]+offset, MAXPASS+1-offset);
+  while (npass > 0)
+    {
+      rbytes = read(fd, passwords[i]+offset, MAXPASS-offset);
```
Looks like commit e89d4c973 is missing in this version.

Can we have a single implementation instead of two different copies, please?

Reply:
Can we use inline functions? I'd like to avoid dependency on libpam in helpers.
Or perhaps have some helper_common.c in modules/ directory.

Reply:
Yes, we can use inline functions.  In fact, we already use them, see e.g. pam_str_skip_prefix_len.
We could introduce a convenience library for use in helpers, but in this case an inline function should be good enough.

Reply:
I've moved read_passwords from pam_unix to pam_inline, which includes [e89d4c9](https://github.com/linux-pam/linux-pam/commit/e89d4c97385ff8180e6e81e84c5aa745daf28a79), and renamed the function to pam_read_passwords.

Reply:
Thanks.  Now I see that MAXPASS macro is defined both in modules/pam_unix/passverify.h and libpam/include/pam_inline.h files.  Could we just drop it and use PAM_MAX_RESP_SIZE instead?

Reply:
Something like commit b36351dd0137034e79194023c8d687b495e785c4 in ldv/next branch.

Reply:
I've rebased my changes on top of that commit, so you'll see it in this PR.

Reply:
Would it be OK for this PR if I push these pam_read_passwords related commits from ldv/next to master?
This PR would have to be rebased to master anyway.

Reply:
Yes, no problem. Just tell me when they are ready.

Reply:
I think they are ready.